fe/fe-core/src/main/java/com/starrocks/transaction/OlapTableTxnLogApplier.java

```diff
@@ -0,0 +1,112 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+package com.starrocks.transaction;
+
+import com.google.common.collect.Lists;
+import com.starrocks.catalog.LocalTablet;
+import com.starrocks.catalog.MaterializedIndex;
+import com.starrocks.catalog.OlapTable;
+import com.starrocks.catalog.Partition;
+import com.starrocks.catalog.Replica;
+import com.starrocks.catalog.Tablet;
+import com.starrocks.sql.optimizer.statistics.IDictManager;
+
+import java.util.List;
+import java.util.Set;
+
+public class OlapTableTxnLogApplier implements TransactionLogApplier {
+    private final OlapTable table;
+
+    public OlapTableTxnLogApplier(OlapTable table) {
+        this.table = table;
+    }
+
+    @Override
+    public void applyCommitLog(TransactionState txnState, TableCommitInfo commitInfo) {
+        Set<Long> errorReplicaIds = txnState.getErrorReplicas();
+        for (PartitionCommitInfo partitionCommitInfo : commitInfo.getIdToPartitionCommitInfo().values()) {
+            long partitionId = partitionCommitInfo.getPartitionId();
+            Partition partition = table.getPartition(partitionId);
+            List<MaterializedIndex> allIndices =
+                    partition.getMaterializedIndices(MaterializedIndex.IndexExtState.ALL);
+            for (MaterializedIndex index : allIndices) {
+                for (Tablet tablet : index.getTablets()) {
+                    for (Replica replica : ((LocalTablet) tablet).getReplicas()) {
+                        if (errorReplicaIds.contains(replica.getId())) {
+                            // should get from transaction state
+                            replica.updateLastFailedVersion(partitionCommitInfo.getVersion());
+                        }
+                    }
+                }
+            }
+            partition.setNextVersion(partition.getNextVersion() + 1);
+        }
+    }
+
+    @Override
+    public void applyVisibleLog(TransactionState txnState, TableCommitInfo commitInfo) {
```

Comment:
Moved from the old method `OlapTableStateMachine.applyVisibleLog` without any change.